contribute-to-tidb/review-a-pr-stage.md

```diff
@@ -0,0 +1,233 @@
+
+* Does the change integrate well with the rest of our system?
+* Is it a good time now to add this functionality?
+
+[TiDB Internals forum](https://internals.tidb.io/) is a place for discussion. For pull requests that are created without prior consensus, try to seek consensus there.
+
+### Functionality
+
+* Does this PR do what the developer intend to do?
+* Is what the developer intend to do good for the users with this implementation?
+
+Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.
+
+You can always validate the PR if you want. It's hard to understand how some changes will impact a user when you're just reading the code, such as a UI change. You can ask the PR author to give you a demo of the functionality if it's too inconvenient to try it yourself.
+
+### Logic
+
+In the case that concurrent programming involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.
+
+The human brain is just like a special machine, it is good at checking a small piece of logic before the states become too complex. Complexity is the biggest challenge of software.
+
+### Complexity
+
+Is the PR more complex than it should be? Check the following items for review the complexity of a PR:
+
+* Are individual lines too complex?
+* Are functions too complex?
+* Are classes too complex?
+
+"Too complex" usually means "can't be understood quickly by readers". It may also mean "developers are likely to introduce bugs when they try to call or modify this code".
+
+A particular type of complexity is over-engineering, where developers have made the code more generic than it needs to be, or added functionality that isn’t presently needed by the system. Reviewers should be especially vigilant about over-engineering. Encourage developers to solve problems must to be solved now, not problems that a developer speculates might need to be solved in the future. 
+
+### Tests
+
+A PR should be test covered, whether the tests are unit tests, integration tests, or end-to-end tests.
+
+Ensure the tests in the PR are correct, sensible, and useful.
+
+* Are they covering interesting cases?
+* Are they readable?
+* Does the PR lower overall test coverage? 
+
+Simulate cases where the code may break.
+
+* Will the tests actually fail when the code is broken?
+* If the code changes beneath them, will they start producing false positives?
+* Does each test makes simple and useful assertions?
+* Are the tests separated appropriately between different test methods?
+
+Remember that tests are also code that have to be maintained, and thus do not accept unnecessary complexity in tests.
```

Comment:
Comment: In some cases, there are two spaces between words. Only use one.

Reply:
Ok. I will do a global search and replace.